apis/python/src/tiledbsoma/soma_array.cc

```diff
@@ -0,0 +1,506 @@
+                        result_order,
+                        timestamp);
+                }),
+            "uri"_a,
+            py::kw_only(),
+            "name"_a = "unnamed",
+            "column_names"_a = py::none(),
+            "batch_size"_a = "auto",
+            "result_order"_a = ResultOrder::automatic,
+            "platform_config"_a = py::dict(),
+            "timestamp"_a = py::none())
+
+        .def(
+            "set_condition", 
+            [](SOMAArray& reader, 
+               py::object py_query_condition,
+               py::object py_schema){
+                   auto column_names = reader.column_names();
+                   // Handle query condition based on
+                   // TileDB-Py::PyQuery::set_attr_cond()
+                   QueryCondition* qc = nullptr;
+                   if (!py_query_condition.is(py::none())) {
+                       py::object init_pyqc = py_query_condition.attr(
+                           "init_query_condition");   
+                       try {
+                           // Column names will be updated with columns present
+                           // in the query condition
+                           auto new_column_names =
+                               init_pyqc(py_schema, column_names)
+                                   .cast<std::vector<std::string>>();   
+                           // Update the column_names list if it was not empty,
+                           // otherwise continue selecting all columns with an
+                           // empty column_names list
+                           if (!column_names.empty()) {
+                               column_names = new_column_names;
+                           }
+                       } catch (const std::exception& e) {
+                           TPY_ERROR_LOC(e.what());
+                       }   
+                       qc = py_query_condition.attr("c_obj")
+                                   .cast<PyQueryCondition>()
+                                   .ptr()
+                                   .get();
+                   }   
+                   reader.reset(column_names);
+
+                    // Release python GIL after we're done accessing python
+                   // objects
+                   py::gil_scoped_release release;   
+                   // Set query condition if present
+                   if (qc) {
+                       reader.set_condition(*qc);
+                   }
+                }, 
+            "py_query_condition"_a,
+            "py_schema"_a)
+
+        .def(
+            "reset",
+            [](SOMAArray& reader,
+               std::optional<std::vector<std::string>> column_names_in,
+               std::string_view batch_size,
+               ResultOrder result_order) {
+                // Handle optional args
+                std::vector<std::string> column_names;
+                if (column_names_in) {
+                    column_names = *column_names_in;
```

Comment:
Do we know that this is always valid? We'll get a runtime_error if the optional is empty.

Reply:
I think this is checking if the optional is empty or not.